### DISCUSSION TITLE
RFC: WebSockets support in Massa API

### DISCUSSION THREAD

**Discussion by aoudiamoncef** [massalabs/massa#3273] [Ideas]
### Problem

Actually, in order to explore Massa blockchain, ours consumers are using short pooling HTTP requests 
which imply some drawbacks:
- Less efficient resources consumption used to handle incoming HTTP requests.
- Retrieved data could be stale or outdated between two calls.
- The temporary data is lost if it's not requested at time.

### Proposition

Massa node tends to produce a consequent amount of data in real time which could be efficiently served over WebSockets. 

The first proposed subscriptions tend to share the state of the Blockchain which was the most common usage from consumers feedbacks.

We'll implement a new JSON-RPC API which supports HTTP + WS. It'll unify the actual private and public API's. It'll be available as an experimental feature in the next release, new ports will be dedicated to it. 

#### Websocket subscriptions

To be implemented:
- subscribe_blocks/unsubscribe_blocks
- subscribe_blocks_headers/unsubscribe_blocks_headers

Candidats:
- subscribe_endorsements/unsubscribe_endorsements
- subscribe_operations/unsubscribe_operations
- subscribe_execution_events/unsubscribe_execution_events

**Technical stack:**
We'll use [jsonrpsee](https://github.com/paritytech/jsonrpsee) crate as a server which supports both HTTP/WebSockets. It offers a bunch of features and tweaks as it’s based on [tokio-tower](https://docs.rs/tower/latest/tower/) and [hyper](https://hyper.rs/) crates. It’s already used in production ready project as [substrate](https://github.com/paritytech/substrate).

**Comment by AurelienFT** [comment]
I think you can add some explanations on how you will fetch the datas from the differents modules. I think using channels could be great.

Also I checked a bit how `jsonrpsee` implemented it and they follow the eth implementation so I think we will have 1 method for the websocket and as first parameter we will specify to which value we ant to subscribe, is that ok ?

**Reply by aoudiamoncef** [reply]
> I think you can add some explanations on how you will fetch the datas from the differents modules. I think using channels could be great.
> 
I'll be back with a complete explanation ASAP. 

> Also I checked a bit how `jsonrpsee` implemented it and they follow the eth implementation so I think we will have 1 method for the websocket and as first parameter we will specify to which value we ant to subscribe, is that ok ?

This is a concret example of a subscription: 

```rust
	/// New produced block.
	#[subscription(
		name = "massa_subscribe_new_blocks" => "massa_new_blocks",
		aliases = ["subscribe_new_blocks", "massa_subscribe_new_blocks"],
		unsubscribe = "massa_unsubscribe_new_blocks",
		unsubscribe_aliases = ["unsubscribe_new_blocks", "massa_unsubscribe_new_blocks"],
		item = Block
	)]
	fn subscribe_new_blocks(&self);
```

**Reply by AurelienFT** [reply]
I think having aliases is a way to lost peoples. We should have only one name for one thing. Also I don't think prefix methods by `massa_` is useful. What the signification of `=>` ?

**Reply by aoudiamoncef** [reply]
Sorry, it was just an example to discover available features.

For us, we should have something like this: 
```rust
	/// New produced block.
	#[subscription(
		name = "subscribe_new_blocks" => "new_blocks",
		unsubscribe = "unsubscribe_new_blocks",
		item = Block
	)]
	fn subscribe_new_blocks(&self);
```

Explanation of `=>` in macro:
```rust
        /// Override the `foo_sub` and use `foo_subNotif` for the notifications.
        ///
        /// The item field indicates which type goes into result field below.
        ///
        /// The notification format:
        ///
        /// ```
        /// {
        ///     "jsonrpc":"2.0",
        ///     "method":"foo_subNotif",
        ///     "params":["subscription":"someID", "result":"some string"]
        /// }
        /// ```
        #[subscription(name = "foo_sub" => "subNotif", unsubscribe = "unsub", item = String)]
        fn sub_override_notif_method(&self);
```

**Reply by AurelienFT** [reply]
It's still unclear to me. `foo_sub` is kind of a namespace but with the name after the `_` reused in the name of the method ? Why there is a need for such a namespace and not only a method name ?

**Reply by aoudiamoncef** [reply]
You could remove `foo` and make it as a namespace in the parent macro. 
It's a small difference in the result(notification) of subscription  : 
```json
{
  "jsonrpc":"2.0",
   "method":"foo_subNotif" instead of "foo_sub", you are controlling the name of this field.
   "params":["subscription":"someID", "result":"some string"]
}
```

Complete examples in crate [documentation](https://docs.rs/jsonrpsee-proc-macros/0.16.1/jsonrpsee_proc_macros/attr.rpc.html)

**Reply by AurelienFT** [reply]
I might be dumb but I don't see why we can't just call the method `subscribe` and having the data we want to subscribe as first parameter ? It's more easy than this namespace thing

**Reply by aoudiamoncef** [reply]
For the Massa implementation, we don't use any namespace. 

So we be KISS as possible. 

This is a concret example of usage from Polkadot dashboard

<img width="1270" alt="image" src="https://user-images.githubusercontent.com/22281426/204776737-93e5b3d0-8445-48fc-8634-7bf95a14ca13.png">

It's a kind of convention to separate subscribe/unsubscribe requests from notifications(responses).

**Reply by AurelienFT** [reply]
Yes ok. Let's see a first example with blocks and blocks headers.
